### PR TITLE
(test) O3-2939: Explicitly break down E2E tests steps in patient-chart  to mimic user experience

### PR DESCRIPTION
**Pull request by kdaud**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [x] My work includes tests or is validated by existing tests.

## Summary
This PR adds changes that are explicitly breaking down patient-chart E2E tests steps to mimic user experience when interacting with the workflows. The tests include;
- Program enrollment test
- Conditions test
- Biometrics test
- Vital test
- Visit test
- Clinical form test

## Related Issue

https://issues.openmrs.org/browse/O3-2939

### CODE REVIEW

**Comment by kdaud** [comment]
@jayasanka-sack @denniskigen could you help review this PR? Thanks.

File: e2e/specs/biometrics.spec.ts [hunk 1]
```diff
@@ -12,25 +12,38 @@ test.beforeEach(async ({ api }) => {
...
     await biometricsPage.goTo(patient.uuid);
   });
 
-  await test.step('And I click the `Record biometrics` link to launch the form', async () => {
+  await test.step('And I click on the `Record biometrics` link to launch the form', async () => {
     await biometricsPage.page.getByText(/record biometrics/i).click();
   });
 
-  await test.step('And then I fill the form', async () => {
+  await test.step('Then I should be redirected to the record vitals and biometrics form', async () => {
```
Comment (denniskigen) [inline e2e/specs/biometrics.spec.ts:26]: Clicking the link launches the form in the workspace on the same page. We're not getting redirected to a different page.

File: e2e/specs/clinical-forms.spec.ts [hunk 2]
```diff
@@ -43,19 +43,31 @@ test('Fill a clinical form', async ({ page, api }) => {
...
   });
 
-  await test.step('And if I fill a form', async () => {
+  await test.step('When I click the `Soap note template` link to launch the form', async () => {
     await chartPage.page.getByText(/soap note template/i).click();
+  });
 
-    await expect(chartPage.page.getByRole('button', { name: /save and close/i })).toBeVisible();
-    await expect(chartPage.page.getByRole('button', { name: /discard/i })).toBeVisible();
+  await test.step('Then I should be redirected to `Soap note template` form', async () => {
```
Comment (denniskigen) [inline e2e/specs/clinical-forms.spec.ts:50]: Same here. Something like: `Then I should see the 'Soap Note template' form in the workspace` sounds preferable to me.

File: e2e/specs/clinical-forms.spec.ts [hunk 3]
```diff
@@ -43,19 +43,31 @@ test('Fill a clinical form', async ({ page, api }) => {
...
 
+  await test.step('When I fill the subjective findings', async () => {
     await chartPage.page.locator('#SOAPSubjectiveFindingsid').fill(subjectiveFindings);
+  });
+
+  await test.step('And I fill the objective findings', async () => {
     await chartPage.page.locator('#SOAPObjectiveFindingsid').fill(objectiveFindings);
+  });
+
+  await test.step('And I fill the assemenent', async () => {
```
Comment (denniskigen) [inline e2e/specs/clinical-forms.spec.ts:62]:
```suggestion
  await test.step('And I fill the "Assessment" question', async () => {
```

File: e2e/specs/clinical-forms.spec.ts [hunk 4]
```diff
@@ -43,19 +43,31 @@ test('Fill a clinical form', async ({ page, api }) => {
...
     await chartPage.page.getByText(/soap note template/i).click();
+  });
 
-    await expect(chartPage.page.getByRole('button', { name: /save and close/i })).toBeVisible();
-    await expect(chartPage.page.getByRole('button', { name: /discard/i })).toBeVisible();
+  await test.step('Then I should be redirected to `Soap note template` form', async () => {
+    await expect(chartPage.page.getByText(/soap note template/i)).toBeVisible();
+  });
 
+  await test.step('When I fill the subjective findings', async () => {
```
Comment (denniskigen) [inline e2e/specs/clinical-forms.spec.ts:54]:
```suggestion
  await test.step('When I fill the "Subjective findings" question', async () => {
```

File: e2e/specs/conditions.spec.ts [hunk 5]
```diff
@@ -22,14 +22,24 @@ test('Record, edit and delete a condition', async ({ page, api }) => {
     await conditionsPage.page.getByText(/record conditions/i).click();
   });
 
-  await test.step('And I fill the form', async () => {
+  await test.step('Then I should be redircted to the conditions form', async () => {
```
Comment (denniskigen) [inline e2e/specs/conditions.spec.ts:25]: Same here.

File: e2e/specs/conditions.spec.ts [hunk 6]
```diff
@@ -22,14 +22,24 @@ test('Record, edit and delete a condition', async ({ page, api }) => {
...
+    await expect(conditionsPage.page.getByText(/record a condition/i)).toBeVisible();
+  });
+
+  await test.step('When I search `Mental status change` as a condition', async () => {
```
Comment (denniskigen) [inline e2e/specs/conditions.spec.ts:29]:
```suggestion
  await test.step('When I search for `Mental status change` in the search box', async () => {
```

File: e2e/specs/conditions.spec.ts [hunk 7]
```diff
@@ -46,20 +56,32 @@ test('Record, edit and delete a condition', async ({ page, api }) => {
     await expect(dataRow).toContainText(/active/i);
   });
 
-  await test.step('Then if I click on the overflow menu and click the `Edit` button', async () => {
-    await expect(conditionsPage.page.getByRole('button', { name: /options/i })).toBeVisible();
-    await conditionsPage.page.getByRole('button', { name: /options/i }).click();
-    await expect(conditionsPage.page.getByRole('menu', { name: /edit or delete condition/i })).toBeVisible();
+  await test.step('When I click the overflow menu of the created condition', async () => {
```
Comment (denniskigen) [inline e2e/specs/conditions.spec.ts:59]:
```suggestion
  await test.step('When I click the overflow menu in the table row with the newly created condition', async () => {
```

File: e2e/specs/conditions.spec.ts [hunk 8]
```diff
@@ -46,20 +56,32 @@ test('Record, edit and delete a condition', async ({ page, api }) => {
...
+      .nth(0)
+      .click();
+  });
+
+  await test.step('And I click on the `Edit` button', async () => {
     await conditionsPage.page.getByRole('menuitem', { name: /edit/i }).click();
   });
 
-  await test.step('And I edit the condition', async () => {
+  await test.step('Then I should be redirected to the conditons form of the created condition`', async () => {
```
Comment (denniskigen) [inline e2e/specs/conditions.spec.ts:70]: Same as above.

File: e2e/specs/conditions.spec.ts [hunk 9]
```diff
@@ -76,16 +98,15 @@ test('Record, edit and delete a condition', async ({ page, api }) => {
...
-  await test.step('And if I click the overflow menu and then click the `Delete` button', async () => {
-    await expect(conditionsPage.page.getByRole('button', { name: /options/i })).toBeVisible();
-    await conditionsPage.page.getByRole('button', { name: /options/i }).click();
-    await expect(conditionsPage.page.getByRole('menu', { name: /edit or delete condition/i })).toBeVisible();
-    await conditionsPage.page.getByRole('menuitem', { name: /delete/i }).click();
-
-    await expect(conditionsPage.page.getByRole('heading', { name: /delete condition/i })).toBeVisible();
-    await expect(conditionsPage.page.getByText(/are you sure you want to delete this condition/i)).toBeVisible();
-    await expect(conditionsPage.page.getByRole('button', { name: /danger delete/i })).toBeVisible();
+  await test.step('When I click the overflow menu of the created condition', async () => {
```
Comment (denniskigen) [inline e2e/specs/conditions.spec.ts:101]: Same as above, if you agree with the suggestion.

File: e2e/specs/program-enrollment.spec.ts [hunk 10]
```diff
@@ -19,26 +19,39 @@ test('Add and edit a program enrollment', async ({ page, api }) => {
...
+    await page.getByText(/record program enrollment/i).click();
   });
 
-  await test.step('And I record a program enrollment', async () => {
-    await programsPage.page.locator('#program').selectOption('64f950e6-1b07-4ac0-8e7e-f3e148f3463f');
-    await programsPage.page.locator('#enrollmentDateInput').fill('04/07/2023');
-    await programsPage.page.locator('#completionDateInput').fill('05/07/2023');
-    await programsPage.page.locator('#completionDateInput').press('Tab');
-    await programsPage.page.locator('#location').selectOption('44c3efb0-2583-4c80-a79e-1f756a03c0a1');
+  await test.step('Then I should be redirected to the `Record program enrollment` form', async () => {
```
Comment (denniskigen) [inline e2e/specs/program-enrollment.spec.ts:25]: Same as above.

File: e2e/specs/program-enrollment.spec.ts [hunk 11]
```diff
@@ -19,26 +19,39 @@ test('Add and edit a program enrollment', async ({ page, api }) => {
...
+    await expect(page.getByText('Record program enrollment', { exact: true })).toBeVisible();
   });
 
-  await test.step('And I click the submit button', async () => {
-    await programsPage.page.getByRole('button', { name: /save and close/i }).click();
+  await test.step('When I select `Hiv care and Treatment`as a program', async () => {
```
Comment (denniskigen) [inline e2e/specs/program-enrollment.spec.ts:29]:
```suggestion
  await test.step('When I select the program named `Hiv Care and Treatment`', async () => {
```

File: e2e/specs/program-enrollment.spec.ts [hunk 12]
```diff
@@ -19,26 +19,39 @@ test('Add and edit a program enrollment', async ({ page, api }) => {
...
+  await test.step('And I set `05/07/2023` as the completion date', async () => {
+    await page.locator('#completionDateInput').fill('05/07/2023');
+    await page.locator('#completionDateInput').press('Tab');
+  });
+
+  await test.step('And I select `Outpatient Clinic` as the enrollment location', async () => {
+    await page.locator('#location').selectOption('44c3efb0-2583-4c80-a79e-1f756a03c0a1');
+  });
+
+  await test.step('And I click on the `save and close` button', async () => {
```
Comment (denniskigen) [inline e2e/specs/program-enrollment.spec.ts:46]:
```suggestion
  await test.step('And I click on the `Save and close` button', async () => {
```

File: e2e/specs/program-enrollment.spec.ts [hunk 13]
```diff
@@ -49,31 +62,43 @@ test('Add and edit a program enrollment', async ({ page, api }) => {
...
   });
 
-  await test.step('And I edit the program enrollment', async () => {
-    await programsPage.page.locator('#enrollmentDateInput').clear();
-    await programsPage.page.locator('#enrollmentDateInput').fill('03/07/2023');
-    await programsPage.page.locator('#completionDateInput').clear();
-    await programsPage.page.locator('#completionDateInput').fill('04/07/2023');
-    await programsPage.page.locator('#completionDateInput').press('Tab');
-    await programsPage.page.locator('#location').selectOption('1ce1b7d4-c865-4178-82b0-5932e51503d6');
+  await test.step('Then I should be redirected to the `Record program enrollment` form of the created program`', async () => {
```
Comment (denniskigen) [inline e2e/specs/program-enrollment.spec.ts:69]: Same as above

File: e2e/specs/record-allergy.spec.ts [hunk 14]
```diff
@@ -18,19 +18,32 @@ test('Record an allergy', async ({ page, api }) => {
     await allergiesPage.goTo(patient.uuid);
   });
 
-  await test.step('And I click the `Record allergy intolerance` link to launch the form', async () => {
+  await test.step('And I click on the `Record allergy intolerance` link to launch the form', async () => {
     await allergiesPage.page.getByText(/record allergy intolerance/i).click();
   });
 
-  await test.step('And I fill the form', async () => {
+  await test.step('Then I should be redirected to the record allergy form', async () => {
```
Comment (denniskigen) [inline e2e/specs/record-allergy.spec.ts:25]: Same as above

File: e2e/specs/vitals.spec.ts [hunk 15]
```diff
@@ -21,21 +21,47 @@ test('Record vital signs', async ({ page, api }) => {
     await vitalsPage.goTo(patient.uuid);
   });
 
-  await test.step('And I click `Record biometrics` link to launch the form', async () => {
+  await test.step('And I click the `Record biometrics` link to launch the form', async () => {
     await vitalsPage.page.getByText(/record vital signs/i).click();
   });
 
-  await test.step('And then I fill the form', async () => {
+  await test.step('Then I should be redirected to the record vitals and biometrics form', async () => {
```
Comment (denniskigen) [inline e2e/specs/vitals.spec.ts:28]: Same as above

File: e2e/specs/vitals.spec.ts [hunk 16]
```diff
@@ -21,21 +21,47 @@ test('Record vital signs', async ({ page, api }) => {
...
+  await test.step('And I add  `37` as the oxygen saturation', async () => {
     await vitalsPage.page.getByRole('spinbutton', { name: /oxygen saturation/i }).fill('98');
+  });
+
+  await test.step('And I add additional notes', async () => {
     await vitalsPage.page.getByPlaceholder(/type any additional notes here/i).fill('Test notes');
   });
 
-  await test.step('And I click the submit button', async () => {
+  await test.step('And I click on the `save and close` button', async () => {
```
Comment (denniskigen) [inline e2e/specs/vitals.spec.ts:64]:
```suggestion
  await test.step('And I click on the `Save and close` button', async () => {
```

**Review by denniskigen — COMMENTED** [review]
Thanks for working on this, @kdaud. I've left a few suggestions.

File: e2e/specs/vitals.spec.ts [hunk 17]
```diff
@@ -21,21 +21,47 @@ test('Record vital signs', async ({ page, api }) => {
...
+
+  await test.step('And I fill `37` as the respiration rate', async () => {
     await vitalsPage.page.getByRole('spinbutton', { name: /respiration rate/i }).fill('16');
+  });
+
+  await test.step('And I fill `37` as the oxygen saturation', async () => {
+    await vitalsPage.page.getByRole('spinbutton', { name: /oxygen saturation/i }).fill('98');
+  });
+
+  await test.step('And I add  `37` as the oxygen saturation', async () => {
```
Comment (denniskigen) [inline e2e/specs/vitals.spec.ts:56]:
```suggestion
  await test.step('And I add `37` as the oxygen saturation', async () => {
```

File: e2e/specs/clinical-forms.spec.ts [hunk 18]
```diff
@@ -43,19 +43,31 @@ test('Fill a clinical form', async ({ page, api }) => {
...
-    await expect(chartPage.page.getByRole('button', { name: /discard/i })).toBeVisible();
+  await test.step('Then I should see the `Soap note template` form launch in the workspace', async () => {
+    await expect(chartPage.page.getByText(/soap note template/i)).toBeVisible();
+  });
 
+  await test.step('When I fill the `Subjective findings` question', async () => {
     await chartPage.page.locator('#SOAPSubjectiveFindingsid').fill(subjectiveFindings);
+  });
+
+  await test.step('And I fill the objective findings', async () => {
```
Comment (denniskigen) [inline e2e/specs/clinical-forms.spec.ts:58]:
```suggestion
  await test.step('And I fill the `Objective findings` question', async () => {
```

File: e2e/specs/clinical-forms.spec.ts [hunk 19]
```diff
@@ -43,19 +43,31 @@ test('Fill a clinical form', async ({ page, api }) => {
...
     await chartPage.page.locator('#SOAPObjectiveFindingsid').fill(objectiveFindings);
+  });
+
+  await test.step('And I fill the `Assessment` question', async () => {
     await chartPage.page.locator('#SOAPAssessmentid').fill(assessment);
+  });
+
+  await test.step('And I fill the plan', async () => {
```
Comment (denniskigen) [inline e2e/specs/clinical-forms.spec.ts:66]:
```suggestion
  await test.step('And I fill the `Plan` question', async () => {
```

File: e2e/specs/conditions.spec.ts [hunk 20]
```diff
@@ -46,20 +56,32 @@ test('Record, edit and delete a condition', async ({ page, api }) => {
...
+      .nth(0)
+      .click();
+  });
+
+  await test.step('And I click on the `Edit` button', async () => {
     await conditionsPage.page.getByRole('menuitem', { name: /edit/i }).click();
   });
 
-  await test.step('And I edit the condition', async () => {
+  await test.step('Then I should see the conditons form of the created condition launch in the workspace`', async () => {
```
Comment (denniskigen) [inline e2e/specs/conditions.spec.ts:70]:
```suggestion
  await test.step('Then I should see the Conditions form launch in the workspace in edit mode', async () => {
```

**Review by jayasanka-sack — APPROVED** [review]
Looks good to me! Thanks @kdaud !

**Pull request merged**